content/en/v3/admin/guides/jenkins/example.md

```diff
@@ -77,39 +80,48 @@ controller:
                       id: "15.7.0"
                       npmPackagesRefreshHours: 72
 ```
-File:`/tmp/customPlugs.yaml` - Plugins
+File: `/tmp/values.yaml.gotmpl` - Plugins
 ```
+# These components will merge with default versionStream/charts/jenkinsci/jenkins/values.yaml.gotmpl
 controller:
   installPlugins:
-    - kubernetes:1.29.0
-    - workflow-aggregator:2.6
-    - git:4.6.0
-    - configuration-as-code:1.47
-    - nodejs:1.3.11
+    - kubernetes:3546.v6103d89542d6
+    - workflow-aggregator:2.7
+    - git:4.11.2
+    - configuration-as-code:1414.v878271fc496f
+    - nodejs:1.5.1
   additionalPlugins:
-    - job-dsl:1.77
-    - github-branch-source:2.9.5
-    - kubernetes-credentials-provider:0.15
-    - jx-resources:1.0.38
+    - jdk-tool:1.5
+    - job-dsl:1.79
+    - github-branch-source:2.10.2
+    - kubernetes-credentials-provider:0.20
+    - tekton-client:1.0.3
+    - pipeline-stage-view:2.24
+  ingress:
+    enabled: true
+    ingressClassName: nginx
+    kubernetes.io/ingress.class: nginx
+    kubernetes.io/tls-acme: true
 ```
+
 ### Initialize Jenkins server configuration
 `jx gitops` commands are used to create the Jenkins server configuration. More information can be found under [Adding Jenkins Server into Jenkins X](/v3/admin/guides/jenkins/getting-started/#adding-jenkins-servers-into-jenkins-x). The example uses the following command sequence:
 ```bash
-cd $JX3HOME/jx3-gke-gsm                           ## cluster repo root
-jx gitops jenkins add --name jx-jenkins           ## add Jenkins
-jx gitops helmfile resolve --namespace jx-jenkins ## resolve charts references (optional but helps later on in editing)
-cp /tmp/values.yaml helmfiles/jx-jenkins          ## Update tool config
-cp /tmp/customPlugs.yaml helmfiles/jx-jenkins     ## Update plugin config
+cd $JX3HOME/jx3-gke-gsm                             ## cluster repo root
+jx gitops jenkins add --name jx-jenkins             ## add Jenkins
+jx gitops helmfile resolve --namespace jx-jenkins   ## resolve charts references (optional but helps later on in editing)
+cp /tmp/values.yaml helmfiles/jx-jenkins            ## Update tool config
+cp /tmp/values.yaml.gotmpl helmfiles/jx-jenkins     ## Update plugin config
 ```
-Modification to the `$JX3HOME/jx3-gke-gsm/helmfiles/jx-jenkins/helmfiles.yaml` file is done to include the additional `customPlugs.yaml`.
+Modification to the `$JX3HOME/jx3-gke-gsm/helmfiles/jx-jenkins/helmfiles.yaml` file is done to include the additional `values.yaml.gotmpl`.
 ```yaml
  - chart: jenkinsci/jenkins
    name: jenkins
    namespace: jx-jenkins
    values:
    - values.yaml
    - ../../versionStream/charts/jenkinsci/jenkins/values.yaml.gotmpl
-   - customPlugs.yaml
+   - values.yaml.gotmpl
```

Comment:
Why this change? Is there some template requirement? or can we use a plain yaml?

Reply:
The change from a yaml to template provides a better method for merging. Actually I wanted to share what I think the logic is so here goes...

Reviewing the Jenkins chart and how we configured it there are **Chart**, **System** (JenkinsX), and **User** level settings. For plugins management there are _installPlugins_ and _additionalPlugins_ config blocks. The Chart generates only installPlugins and the definiton is:

```
  installPlugins:
    - kubernetes:1.31.3
    - workflow-aggregator:2.6
    - git:4.10.2
    - configuration-as-code:1414.v878271fc496f

```
The System generates only additionalPlugins (inherting the installPlugins from Chart) and the definition is:

```
  additionalPlugins:
  - job-dsl:1.77
  - github-branch-source:2.10.2
  - kubernetes-credentials-provider:0.16
  - jx-resources:1.0.38
  - tekton-client:1.0-alpha3+165
```


If the User needs to add, delete, or update a plugin it will have to be determine whether to include either of the two blocks above. If the requirement is to change or remove an existing plugin you should include the entire block with the change. For example to update the kubernetes plugin and to remove the configuration-as-code plugin you would include the following installPlugins block:

```
  installPlugins:
    - kubernetes:1.31.99                       ## Updated plugin
    - workflow-aggregator:2.6
    - git:4.10.2
#   - configuration-as-code:1414.v878271fc496f ## Removed plugin

```
To add or modify  plugins you would include the following additionalPlugins block:

```
  additionalPlugins:
  - job-dsl:1.79                    ## Modified
  - github-branch-source:2.10.2
  - kubernetes-credentials-provider:0.16
  - jx-resources:1.0.38
  - tekton-client:1.0-alpha3+165
  - jdk-tool:1.5                    ## Added
```

It may be possible to update and "installPlugins" entry by adding it into the "additionalPlugins" list but it has not been tested and it is not recommended.

I can modify this current documentation example to include this above logic. The new version will not have any installPlugins because it will be picked up from the Chart. It will have an enhanced additionalPlugins block which will include plugins for example purposes and the addition for the ingress so that it works. We still have to update verstionStream template to fix the problem but the example doc will still be in good shape. Let me know.

Reply:
I'm totally with you that the merging logic you describe is the standard one for helm. (helm/helm#3486 is an old and closed issue about it.)

But you seem to imply that with the help of helmfile it works differently for gotmpl files. Do you have any reference for that?

Reply:
Don't have a reference. I have been working trying different configurations to see what works. I switched from yaml to template to attempt to have a consistent file naming convention that hopefully that makes the documentation  easy to understand. I thought that using a template might make extending the functionality easier.

Reply:
I did some more testing and it looks like you can implement Jenkins in fewer steps than I had previously thought. It can be done with the modification of the values.yaml file only which means that you don't have to have an additional template file (values.yaml.gotmpl) or make additional modifications to the helmfile.yaml. I've updated the document version to reflect this new config. 

In a nutshell you can initalize Jenkins
```
jx gitops jenkins add --name jx-jenkins             ## add Jenkins
jx gitops helmfile resolve --namespace jx-jenkins   ## resolve charts references (optional but helps later on in editing)
```
and update repohome/helmfiles/jx-jenkins/values.yaml with the following:
```
controller:
  ingress:             ## Required versionStream DNS/Ingress break fix
    ingressClassName: nginx
```
This will use all defaults from the Chart and the current version stream template.

If you need to do specific updates to the Jenkins environment you then possibly need to implement additional files and modification to the helmfile.yaml.